python/ray/rllib/ddpg2/models.py

```diff
@@ -0,0 +1,408 @@
+            output_actions = _build_action_network(
+                self.p_t,
+                low_action,
+                high_action,
+                deterministic_flag,
+                zero_eps,
+                config["exploration_theta"],
+                config["exploration_sigma"])
+
+            output_actions_estimated = _build_action_network(
+                self.p_tp1,
+                low_action,
+                high_action,
+                deterministic_flag,
+                zero_eps,
+                config["exploration_theta"],
+                config["exploration_sigma"])
+
+        # q network evaluation
+        with tf.variable_scope("q_func", reuse=True) as scope:
+            self.q_t = _build_q_network(registry, s_values, act_t, config)
+        with tf.variable_scope("q_func", reuse=True) as scope:
+            self.q_tp0 = _build_q_network(registry, s_values, output_actions, config)
+
+        # target q network evalution
+        with tf.variable_scope("target_q_func") as scope:
+            self.q_tp1 = _build_q_network(registry, target_s_values, output_actions_estimated, config)
+            self.target_q_func_vars = _scope_vars(scope.name)
+
+        q_t_selected = tf.squeeze(self.q_t)
+
+        q_tp1_best = tf.squeeze(input=self.q_tp1)
+        q_tp1_best_masked = (1.0 - done_mask) * q_tp1_best
+
+        # compute RHS of bellman equation
+        q_t_selected_target = (rew_t + config["gamma"] ** config["n_step"] * q_tp1_best_masked)
+
+        # compute the error (potentially clipped)
+        self.td_error = q_t_selected - tf.stop_gradient(q_t_selected_target)
+        errors = _huber_loss(self.td_error)
+
+        weighted_error = tf.reduce_mean(importance_weights * errors)
+
+        self.loss = weighted_error
+
+        # for policy gradient
+        self.actor_loss = -1.0 * tf.reduce_mean(self.q_tp0)
+
+
+class DDPGGraph(object):
+    def __init__(self, registry, env, config, logdir):
+        self.env = env
+        dim_actions = env.action_space.shape[0]
+        low_action = env.action_space.low
+        high_action = env.action_space.high
+        actor_optimizer = tf.train.AdamOptimizer(learning_rate=config["actor_lr"])
+        critic_optimizer = tf.train.AdamOptimizer(learning_rate=config["critic_lr"])
+
+        # Action inputs
+        self.stochastic = tf.placeholder(tf.bool, (), name="stochastic")
+        self.eps = tf.placeholder(tf.float32, (), name="eps")
+        self.cur_observations = tf.placeholder(
+            tf.float32, shape=(None,) + env.observation_space.shape)
+
+        # State outputs
+        s_scope_name = TOWER_SCOPE_NAME + "/s_func"
+        with tf.variable_scope(s_scope_name) as scope:
+            s_values = _build_s_network(
+                registry, self.cur_observations, config)
+            s_func_vars = _scope_vars(scope.name)
+
+        # Actor: P (policy) network
+        p_scope_name = TOWER_SCOPE_NAME + "/p_func"
```

Comment:
We can drop TOWER SCOPE name, that was only used for the multi GPU optimizer, which is being refactored to not need it